engine/execution/computation/computer/uploader/uploader.go

```diff
@@ -0,0 +1,151 @@
+package uploader
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/module"
+	"github.com/rs/zerolog"
+
+	"github.com/sethvargo/go-retry"
+)
+
+type Uploader interface {
+	Upload(computationResult *execution.ComputationResult) error
+}
+
+func NewAsyncUploader(uploader Uploader, retryInitialTimeout time.Duration, maxRetryNumber uint64, log zerolog.Logger, metrics module.ExecutionMetrics) *AsyncUploader {
+	return &AsyncUploader{
+		unit:                engine.NewUnit(),
+		uploader:            uploader,
+		log:                 log.With().Str("component", "block_data_uploader").Logger(),
+		metrics:             metrics,
+		retryInitialTimeout: retryInitialTimeout,
+		maxRetryNumber:      maxRetryNumber,
+	}
+}
+
+type AsyncUploader struct {
+	unit                *engine.Unit
+	uploader            Uploader
+	log                 zerolog.Logger
+	metrics             module.ExecutionMetrics
+	retryInitialTimeout time.Duration
+	maxRetryNumber      uint64
+}
+
+func (a *AsyncUploader) Ready() <-chan struct{} {
+	return a.unit.Ready()
+}
+
+func (a *AsyncUploader) Done() <-chan struct{} {
+	return a.unit.Done()
+}
+
+func (a *AsyncUploader) Upload(computationResult *execution.ComputationResult) error {
+
+	fibRetry, err := retry.NewFibonacci(a.retryInitialTimeout)
+	if err != nil {
+		return fmt.Errorf("cannot create retry mechanism: %w", err)
+	}
+	cappedFibRetry := retry.WithMaxRetries(a.maxRetryNumber, fibRetry)
+
+	a.unit.Launch(func() {
+		a.metrics.ExecutionBlockDataUploadStarted()
+		start := time.Now()
+
+		err := retry.Do(a.unit.Ctx(), cappedFibRetry, func(ctx context.Context) error {
+			err := a.uploader.Upload(computationResult)
+			if err != nil {
+				a.log.Warn().Err(err).Msg("error while uploading block data, retrying")
+			}
+			return retry.RetryableError(err)
+		})
+
+		if err != nil {
+			a.log.Error().Err(err).Msg("failed to upload block data")
```

Comment:
```suggestion
			a.log.Error().Err(err).Str("block_id", computationResult.ExecutableBlock.Block.ID()).Msg("failed to upload block data")
```

Just so we know which one failed if we need to manually fix it later